### ISSUE TITLE
Code Example: Using Google Maps' Places Autocomplete API for a set of Address Fields in Pods

### ISSUE THREAD

**Issue by leepeterson** [pods-framework/pods-docs#4]
## The Premise

You're developing a website with a content type made for physical, geographic locations. Each location created requires basic business info like Street Address, City, State, as well as Phone, Website, Reviews. Sure, you could enter in each location's address and contact info one field at a time, one location at a time, but how do you know if what you're entering is up-to-date and accurate? And do you really want to enter all these one at a time? No, of course not.

This is where Google Maps' Places API comes into play. And to make things even easier, there are a multitude of JavaScript-based libraries made to connect Pods-based input fields with the Places Geocoding API that offer autocomplete functionality and provide suggested locations as you begin typing an address.

## What You'll Need

• [Pods v2.7+](http://wordpress.org/plugins/pods/)
• [Google Maps API Key](https://cloud.google.com/maps-platform)
• [Geocomplete](https://github.com/ubilabs/geocomplete) - One such option out of many

## How to Approach the Task

1. Create a meta group containing several text fields, one of which is used solely to type an address into and is also the one offering the list of suggested locations pulled from Google Maps' Places API
2. Assign the meta group to location-based content types within WordPress
3. Load the Google Places API script only on post edit screens
4. Write a bit of JavaScript that listens on the one field you'd like to offer suggestions from
5. Ensure fields are created for the various parts of a location's address so they're filled in when a suggested address is chosen

## Creating a Pods Meta Group

This assumes you've already created a content type using Pods. 

To help you determine which CPTs can be classified as a "location", you could create a global helper function that lists a static array of CPT names:

```php
function get_location_types() {

	static $location_types = [ 'restaurant', 'hotel', 'park' ];

	return $location_types;

}
```

Since you only want to enqueue the Places API on "location" post edit screens, you'll want to create a function that determines when a "location" post edit screen is being viewed:

```php
function is_location_type() {

	global $post, $typenow, $current_screen;
	
	$type  = '';
	$location_types = get_location_types();

	if ( $post && $post->post_type ) {
		$type = $post->post_type;
	}
	elseif ( $typenow ) {
		$type = $typenow;
	}
	elseif ( $current_screen && $current_screen->post_type ) {
		$type = $current_screen->post_type;
	}
	elseif ( isset( $_REQUEST['post_type'] ) ) {
		$type = sanitize_key( $_REQUEST['post_type'] );
	}
	elseif ( isset( $_REQUEST['post'] ) ) {
		$type = get_post_type( $_REQUEST['post'] );
	}

	if ( in_array( $type, $types ) ) 
		return true;

}
```

Write a function that holds onto the enqueue functions:

```php
function location_enqueue_scripts() {

	wp_register_script( 'admin/location/geocode', getenv('GEOCODE_API'), false, null, true );
	wp_register_script( 'admin/location/plugins', getenv('STATIC_CDN').'/js/location-plugins.js', ['jc/admin/geocode'], null, true );
	wp_enqueue_script( 'admin/location/init', getenv('STATIC_CDN').'/js/location-init.js', ['jc/admin/plugins'], null );

}
```

Create a meta group containing your address fields, paying special attention to the field that is listened on (in this case, 'address_lookup')

```php
add_action( 'pods_meta_groups', function( $type, $name ) {
	$is_location_type = is_location_type();
	$location_types = get_location_types();

	$location_address = [
		'address_lookup' => [
			'label'   => 'Address Lookup',
			'type'    => 'text',
			'attributes' => [
				'id'          => 'address-lookup-field',
				'placeholder' => 'Begin typing an address&hellip;',
			],
			'description' => '<div id="address-lookup-map" style="width:95%;border:1px solid transparent;background:#fff"></div>',
		],
		'location_name' => [
			'label'   => 'Location Name',
			'type'    => 'text',
			'attributes' => [
				'data-geo' => 'name',
			]
		],
		'full_address' => [
			'label'   => 'Full Address',
			'type'    => 'text',
			'attributes' => [
				'data-geo' => 'formatted_address',
			]
		],
		'street_number' => [
			'label'   => 'Street Number',
			'type'    => 'text',
			'attributes' => [
				'data-geo' => 'street_number',
			]
		],
		'street_name' => [
			'label'   => 'Street Name',
			'type'    => 'text',
			'attributes' => [
				'data-geo' => 'route',
			]
		],
		'city' => [
			'label'   => 'City',
			'type'    => 'text',
			'attributes' => [
				'data-geo' => 'locality',
			]
		],
		'state' => [
			'label'   => 'State',
			'type'    => 'text',
			'attributes' => [
				'data-geo' => 'administrative_area_level_1',
			]
		],
		'zip_code' => [
			'label'   => 'Zip Code',
			'type'    => 'number',
			'attributes' => [
				'data-geo' => 'postal_code',
			]
		],
		'country' => [
			'label'   => 'Country',
			'type'    => 'text',
			'attributes' => [
				'data-geo' => 'country',
			]
		],
		'latitude' => [
			'label'   => 'Latitude',
			'type'    => 'number',
			'attributes' => [
				'data-geo' => 'lat',
			]
		],
		'longitude' => [
			'label'   => 'Longitude',
			'type'    => 'number',
			'attributes' => [
				'data-geo' => 'lng',
			]
		]
	];
	pods_group_add( $location_types, 'Location Address', $location_address, 'normal', 'core', 'post_type' );

	if ( true === $is_location_type ) {
		add_action( 'admin_enqueue_scripts', 'location_enqueue_scripts' );
	}

}, 10, 2 );
```

What we've done above is we've created the Pods fields in CPTs classified as "location" types, and we've enqueued the Places API, only on these specific CPT edit screens.

As noted in "What you'll need", we're using the jQuery library Geocode to help interface with the Google Maps API. All that's left now is a bit of jQuery to finish the connection between the Places API and the field being typed into:

```javascript
window.onload = function() {

	var geo_map = document.getElementById('address-lookup-map'),
		geo_field = document.getElementById('address-lookup-field'),
		geo_form = document.getElementById('post');

	jQuery(geo_field).geocomplete({
		map: geo_map,
		details: geo_form,
		detailsAttribute: 'data-geo',
		types: ['geocode', 'establishment']
	}).bind('geocode:result', function(event, result) {
		jQuery(geo_map).css('height', '20em').css('border-color', '#ddd');
	});

};
```

## In Conclusion

Sure, there are several APIs available to obtain location information, and many solutions to integrate with a user interface. But this is just one such example, probably one of the quickest and easiest to maintain long term. 

Further documentation on the Geocomplete jQuery plugin can be found [here](http://ubilabs.github.com/geocomplete/), and, while this plugin is labeled "archived" and "not maintained", it works well enough for something as simple as this example.
